Bioconductor scater scripts

suhaibMo opened the pull request:
FOR CONTRIBUTOR:
* [ ] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [ ] - License permits unrestricted use (educational + commercial)
* [ ] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

FOR REVIEWER:
* [ ] .shed.yml file ok
    - [ ] Toolshed user `iuc` has access to associated toolshed repo(s)
* [ ] Indentation is correct (4 spaces)
* [ ] Tool version/build ok
* [ ] `<command/>`
  - [ ] Text parameters, input and output files `'single quoted'`
  - [ ] Use of `<![CDATA[ ... ]]>` tags
  - [ ] Parameters of type `text` or having `optional="true"` attribute are checked with `if str($param)` before being used
* [ ] Data parameters have a `format` attribute containing datatypes recognised by Galaxy
* [ ] Tests
  - [ ] Parameters are reasonably covered
  - [ ] Test files are appropriate
* [ ] Help
  - [ ] Valid restructuredText and uses `<![CDATA[ ... ]]>` tags
* [ ] Complies with other best practice in [Best Practices Doc](http://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html)

pcm32 commented on tools/scater_scripts/scater-calculate-cpm.xml:
```diff
@@ -0,0 +1,51 @@
+        						-t "$cpm_matrix"
+    ]]></command>
+    <inputs>
+        <param type="data" name="R_scater_serialized" format="rds" />
```
More of a question for Galaxy devs: Should we use rds or Rdata directly? I have personally had issues with rds, but if this is the current galaxy way to go...

pcm32 replied:
@suhaibMo at other places you have rdata, so i would go with that I guess instead of rds.

mblue9 replied:
rds doesn't currently exist as a Galaxy datatype so you'll have to use rdata

pcm32 commented on tools/scater_scripts/scater-calculate-qc-metrics.xml:
```diff
@@ -0,0 +1,77 @@
+        								-o "$R_scater_qc" 
+    ]]></command>
+    <inputs>
+        <param type="data" name="R_scater_serialized" format="rds" />
```
rds/rdata ?

pcm32 commented on tools/scater_scripts/scater-normalize.xml:
```diff
@@ -0,0 +1,75 @@
+        </test>
+    </tests>
+    <help><![CDATA[
+	Usage: /Users/suhaib/miniconda/bin/scater-normalize.R [options]
```
Maybe remove the path here? specially since it is a path in your own machine :-)

pcm32 commented:
LGTM @suhaibMo! Great work!

bgruening commented on tools/scater_scripts/scater-calculate-cpm.xml:
```diff
@@ -0,0 +1,51 @@
+    <command detect_errors="exit_code"><![CDATA[
+        ln -s '$R_scater_serialized' R_scater_serialized.rds;
+        
+        scater-calculate-cpm.R -i "$R_scater_serialized" 
```
please use single quotes, double quotes gets evaluated by bash and can be a security issue.
See here for more information: https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#command-formatting

bgruening commented on tools/scater_scripts/scater-calculate-cpm.xml:
```diff
@@ -0,0 +1,51 @@
+		File name in which a serialized R SingleCellExperiment object where object matrix found  
+
+	-s SIZE-FACTORS, --size-factors=SIZE-FACTORS
+		A  logical  scalar  indicating  whether  size  factors  in object should  be  used  to compute effective library sizes.  If not, all size factors are deleted and librarysize-based factors are used instead (seelibrarySizeFactors.  Alternatively, anumeric vector containing a size factor for each cell, which is used in place ofsizeFactor(object).
```
This information should ideally go into the parameter description on top and removed here.

bgruening commented on tools/scater_scripts/scater-calculate-cpm.xml:
```diff
@@ -0,0 +1,51 @@
+    ]]></command>
+    <inputs>
+        <param type="data" name="R_scater_serialized" format="rds" />
+		<param name="size_factors" type="boolean" label= "use size factors"
```
you could use `argument` here: https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#parameter-name-argument-and-help

bgruening commented on tools/scater_scripts/scater-calculate-cpm.xml:
```diff
@@ -0,0 +1,51 @@
+    <inputs>
+        <param type="data" name="R_scater_serialized" format="rds" />
+		<param name="size_factors" type="boolean" label= "use size factors"
+       			help="(-s)" optional="True" />
```
optional does not make much sense in a boolean imho.

bgruening commented on tools/scater_scripts/scater-calculate-qc-metrics.xml:
```diff
@@ -0,0 +1,77 @@
+<tool id="scater-calculate-qc-metrics" name="Scater calculate qc metrics" version="0.1.0">
```
You have redundant information in your `name` and `description`, e.g. `calculate qc metrics`

bgruening commented on tools/scater_scripts/scater-normalize.xml:
```diff
@@ -0,0 +1,75 @@
+    ]]></command>
+    <inputs>
+        <param type="data" name="R_scater_qc" format="rds" />
+        <param type="text" name="exprs_value" optional="True" label="exprs values" help="Default: counts" />
```
What are valid inputs here? Should this be sanitized? If the default is `counts` you can add it as `value`. But maybe this could be a select field?

bgruening replied:
The label should also be a little bit more expressive I think.

bgruening commented on tools/scater_scripts/scater-read-10x-results.xml:
```diff
@@ -0,0 +1,48 @@
+    ]]></command>
+    <inputs>
+        <param type="data" name="matrix" format="mtx" />
+        <param type="data" name="genes" format="tsv" />
```
please use `tabular` if you are referring to a tabular file.

bgruening commented on tools/scater_scripts/scater-read-10x-results.xml:
```diff
@@ -0,0 +1,48 @@
+<tool id="scater-read-10x-results" name="Scater read 10x data" version="0.1.0">
```
the tool-ids should probably reflect the scater version.

suhaibMo replied:
@bgruening The tool-id I understand `is a short identifier used by Galaxy itself to identify the tool` https://planemo.readthedocs.io/en/latest/writing_standalone.html
since we are  bioconda packaged tool to call the scripts, we are using the script name to identify.

bgruening replied:
@suhaibMo see here: https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#tool-versions

suhaibMo replied:
Thanks @bgruening

pcm32 replied:
Shouldn't in this case maybe @bgruening reflect the version of the intermediate bioconda package? Or should we add that as a secondary version? after a dash or something maybe?

nsoranzo replied:
I'd use the version of the bioconductor-scater-scripts package.

suhaibMo replied:
ok. considering I'm providing version of bioconductor-scater-scripts package in bioconda in the macros 
`<requirement type="package" version="0.0.3">bioconductor-scater-scripts</requirement>`

I will add something like to reflect tool id as 
1. `<tool id="scater-read-10x-results-0.0.3" name="Scater read 10x data" version="0.0.3"` ? or 
2. `<tool id="scater-read-10x-results" name="Scater read 10x data" version="0.0.3">`

The version of scater (in bioconductor) is 1.6 which is in macros as well
`<xml name="version">
      <version_command><![CDATA[
echo $(R --version | grep version | grep -v GNU)", scater version" $(R --vanilla --slave -e "library(scater); cat(sessionInfo()\$otherPkgs\$scater\$Version)" 2> /dev/null | grep -v -i "WARNING: ")
    ]]></version_command>`

Let me know @nsoranzo @bgruening option 1 or 2 ?

bgruening replied:
2) :) Maybe we can actually do both by adding version 1.6 to the bioconda-scater-scripts ... then everything would be consistent. The id should identify this tool - not the version. In Galaxy would want to have an upgrade path between different versions of a particular tool - so the ID needs to stay the same.

bgruening commented on tools/scater_scripts/scater_macros.xml:
```diff
@@ -0,0 +1,22 @@
+<macros>
+    <xml name="requirements">
+        <requirements>
+        <requirement type="package" version="0.0.3">bioconductor-scater-scripts</requirement>
```
Since we are not pinning the scater version in the bioconda recipe, we should probably put in here a supported scater vresion.

suhaibMo commented:
Thank you @pcm32  @bgruening @mblue9  for your feedback and suggestions. Much appreciated.

suhaibMo commented:
I've tried to address PR suggestions as much as possible. Now most of the  functional wrapper are added with tests. Thank you @bgruening @mblue9 @pcm32

suhaibMo commented:
Hey @pcm32 @bgruening @mblue9 , this PR been around for a while. I've addressed the comments and suggestions provided to improve. Can this be looked into to merge please.

pcm32 commented:
@suhaibMo not sure if the `<help>` sections are adhering to best practices for instance. They look like cli instructions, which shouldn't be for Galaxy.

nsoranzo commented:
> @suhaibMo not sure if the `<help>` sections are adhering to best practices for instance. They look like cli instructions, which shouldn't be for Galaxy.

For guidelines see https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#tool-versions

suhaibMo commented:
linting check fails by moving help sections `Element 'data', attribute 'help': The attribute 'help' is not allowed`. I'm moving the help sections from parameters that has type data.

suhaibMo commented:
Can anyone point the reason CI is failing ? Lint check its passes without any errors on my personal computer.

nsoranzo commented:
@suhaibMo See https://github.com/galaxyproject/tools-iuc/pull/2160

suhaibMo commented:
@nsoranzo thanks. I'm getting error W605 in addition to W503 and W504. Does this call to ignore these errors until https://github.com/galaxyproject/tools-iuc/pull/2160 is fixed. In that case, considering this PR is adhering to best practices now, can someone vouch to approve please.

pcm32 commented:
@suhaibMo I don't think that you should have removed the help attribute from the inputs. It was correctly removed for the outputs, but needs to be restored for inputs. The message of the linter that you mentioned regarding help fields is about the `<data >` entries for the outputs, it doesn't apply to the inputs.

suhaibMo commented:
@pcm32 my bad, the input help attributes are restored.

bernt-matthias commented:
Seems that scater is now already available https://github.com/galaxyproject/tools-iuc/tree/master/tools/scater
